### PR TITLE
feat: add CORS headers to responses and an OPTIONS handler

### LINKED ISSUE #47: Set CORS headers on responses
Currently we're not settings CORS headers on any of the responses, because I made the incorrect assumption when I set this up that the endpoints wouldn't be called from a browser.

We can copy over the stuff from we3.storage, see:
* [packages/api/src/cors.js](https://github.com/web3-storage/web3.storage/blob/main/packages/api/src/cors.js)
* [packages/api/src/index.js#L32-L60](https://github.com/web3-storage/web3.storage/blob/main/packages/api/src/index.js#L32-L60)

### PR DESCRIPTION
**Pull request by adamalton**
Fixes #47

### CODE REVIEW

File: packages/api/src/cors.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    const respHeaders = {
+      'Content-Length': '0',
+      'Access-Control-Allow-Origin': headerOrDefault('origin', '*'),
+      'Access-Control-Allow-Methods': 'GET,POST,DELETE,OPTIONS',
+      'Access-Control-Max-Age': '86400',
+      // Allow all future content Request headers to go back to browser
+      // such as Authorization (Bearer) or X-Client-Name-Version
+      'Access-Control-Allow-Headers':
+        headerOrDefault('Access-Control-Request-Headers', ''),
+      'Access-Control-Expose-Headers': 'Link, Count, Page, Size'
```
Comment (joshJarr) [inline packages/api/src/cors.ts:33]: I don't think we need these as `Link, Count, Page, Size'` are all for pagination which w3name doesn't use

File: packages/api/src/cors.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  // Clone the response so that it's no longer immutable (like if it comes from cache or fetch)
+  response = new Response(response.body, response)
+  const origin = request.headers.get('origin')
+  if (origin !== '' && origin !== null) {
+    response.headers.set('Access-Control-Allow-Origin', origin)
+    response.headers.set('Vary', 'Origin')
+  } else {
+    response.headers.set('Access-Control-Allow-Origin', '*')
+  }
+  response.headers.set('Access-Control-Expose-Headers', 'Link, Count, Page, Size')
```
Comment (joshJarr) [inline packages/api/src/cors.ts:71]: We can also remove this Access-Control-Expose-Headers

**Review by joshJarr — CHANGES_REQUESTED** [review]
LGTM, might be good to remove these unused `Access-Control-Expose-Headers` pagination headers as we don't use 'em in w3name. Thanks!

File: packages/api/src/cors.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        return value
+      }
+      return fallback
+    }
+    // Handle CORS pre-flight request.
+    /** @type {Record<string, string>} */
+    const respHeaders = {
+      'Content-Length': '0',
+      'Access-Control-Allow-Origin': headerOrDefault('origin', '*'),
+      'Access-Control-Allow-Methods': 'GET,POST,DELETE,OPTIONS',
```
Comment (joshJarr) [inline packages/api/src/cors.ts:27]: I don't think we have any DELETE endpoints? So could omit this for now tooo

File: packages/api/test/cors.spec.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      )
+    )
+    // const res = await fetch(new URL('/name/abcd', endpoint), {
+    //   method: 'OPTIONS',
+    //   headers: {
+    //     Origin: 'name.web3.storage',
+    //     'Access-Control-Request-Method': 'whatever',
+    //     'Access-Control-Request-Headers': 'whatever'
+    //   }
+    // })
```
Comment (joshJarr) [inline packages/api/test/cors.spec.ts:61]: Looks like an outdated comment, did you want to keep the commented `fetch()` examples in these tests?

File: packages/api/src/cors.ts (hunk 1 above)
Comment (adamalton) [reply packages/api/src/cors.ts:33]: Good point.

File: packages/api/src/cors.ts (hunk 2 above)
Comment (adamalton) [reply packages/api/src/cors.ts:71]: Also a good point.

File: packages/api/src/cors.ts (hunk 3 above)
Comment (adamalton) [reply packages/api/src/cors.ts:27]: Again, a good point.

File: packages/api/test/cors.spec.ts (hunk 4 above)
Comment (adamalton) [reply packages/api/test/cors.spec.ts:61]: Ah yes, I had to convert from using `fetch` to using `mf.dispatchFetch`, and this just got left behind. Will remove.

**Comment by adamalton** [comment]
@joshJarr I've made changes as per your comments. Ready for re-review.

**Pull request merged**